ArmPkg/Library/ArmFfaLib/ArmFfaCommon.c

```diff
@@ -0,0 +1,758 @@
+    FfaArgs->Arg2 = ARM_FFA_RET_NOT_SUPPORTED;
+    return TRUE;
+  }
+
+  return FALSE;
+}
+
+/**
+  Check FF-A support or not.
+
+  @retval TRUE                   Supported
+  @retval FALSE                  Not supported
+
+**/
+BOOLEAN
+EFIAPI
+IsFfaSupported (
+  IN VOID
+  )
+{
+  return gFfaSupported;
+}
+
+/**
+  Get FF-A version.
+
+  @param [in]    RequestMajorVersion          Minimal request major version
+  @param [in]    RequestMinorVersion          Minimal request minor version
+  @param [out]   CurrentMajorVersion          Current major version
+  @param [out]   CurrentMinorVersion          Current minor version
+
+**/
+EFI_STATUS
+EFIAPI
+ArmFfaLibVersion (
+  IN  UINT16  RequestMajorVersion,
+  IN  UINT16  RequestMinorVersion,
+  OUT UINT16  *CurrentMajorVersion,
+  OUT UINT16  *CurrentMinorVersion
+  )
+{
+  ARM_FFA_ARGS  FfaArgs;
+
+  ZeroMem (&FfaArgs, sizeof (ARM_FFA_ARGS));
+
+  FfaArgs.Arg0 = ARM_FID_FFA_VERSION;
+  FfaArgs.Arg1 = ARM_FFA_CREATE_VERSION (
+                   RequestMajorVersion,
+                   RequestMinorVersion
+                   );
+
+  ArmCallFfa (&FfaArgs);
+
+  if (IsFfaError (&FfaArgs)) {
+    return FfaStatusToEfiStatus (FfaArgs.Arg2);
+  }
+
+  if (CurrentMajorVersion != NULL) {
+    *CurrentMajorVersion = ARM_FFA_MAJOR_VERSION_GET (FfaArgs.Arg0);
+  }
+
+  if (CurrentMinorVersion != NULL) {
+    *CurrentMinorVersion = ARM_FFA_MINOR_VERSION_GET (FfaArgs.Arg0);
+  }
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Get FF-A features.
+
+  @param [in]   Id               Feature id or function id
+  @param [in]   InputProperties  Input properties according to Id
+  @param [out]  Property1        First property.
+  @param [out]  Property2        Second property.
+
+  @retval EFI_SUCCESS
+  @retval Others                 Error
+
+**/
+EFI_STATUS
+EFIAPI
+ArmFfaLibFeatures (
+  IN  UINT32  Id,
+  IN  UINT32  InputProperties,
+  OUT UINTN   *Property1,
+  OUT UINTN   *Property2
+  )
+{
+  ARM_FFA_ARGS  FfaArgs;
+
+  if ((Property1 == NULL) || (Property2 == NULL)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  ZeroMem (&FfaArgs, sizeof (ARM_FFA_ARGS));
+  *Property1 = 0x00;
+  *Property2 = 0x00;
+
+  switch (Id) {
+    case ARM_FID_FFA_RXTX_MAP_AARCH32:
+    case ARM_FID_FFA_RXTX_MAP_AARCH64:
+      if ((InputProperties != 0x00)) {
+        DEBUG ((DEBUG_ERROR, "%a: Invalid Parameter for FunctionId: 0x%x", __func__, Id));
+        return EFI_INVALID_PARAMETER;
+      }
+
+      break;
+  }
+
+  FfaArgs.Arg0 = ARM_FID_FFA_FEATURES;
+  FfaArgs.Arg1 = Id;
+  FfaArgs.Arg2 = InputProperties;
+
+  ArmCallFfa (&FfaArgs);
+
+  if (IsFfaError (&FfaArgs)) {
+    return FfaStatusToEfiStatus (FfaArgs.Arg2);
+  }
+
+  switch (Id) {
+    case ARM_FID_FFA_RXTX_MAP_AARCH32:
+    case ARM_FID_FFA_RXTX_MAP_AARCH64:
+    case ARM_FFA_FEATURE_ID_NOTIFICATION_PENDING_INTERRUPT:
+    case ARM_FFA_FEATURE_ID_SCHEDULE_RECEIVER_INTERRUPT:
+    case ARM_FFA_FEATURE_ID_MANAGED_EXIT_INTERRUPT:
+      *Property1 = FfaArgs.Arg2;

```

Comment:
Maybe I missed something, but where is the property 2 being used here?

Also, for my own education, how does other function id query respond in this case? Just EFI_SUCCESS?

Reply:
>Maybe I missed something, but where is the property 2 being used here?

Currently, There is no usecase. but when you see the Table 13.14,
Return Parameters can have 2 (w2-w3). so I've added property2 in here.

> Also, for my own education, how does other function id query respond in this case? Just EFI_SUCCESS?

Yes. If function id supports it returns FFA_SUCCESS so it's mapped to return EFI_SUCCESS because wh